rfcs/2020-09-28-cloudwatch-metrics.md

```diff
@@ -0,0 +1,181 @@
+aws_ec2_cpu_credit_balance_sample_count{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 1
+aws_ec2_cpu_credit_balance_sum{instance_id="i-0db9620c0ee32d463",region="us-east-1"} 576
+```
+
+I think we will want to provide example configs that collect common metrics for
+various namespaces (like EC2, S3, etc.).
+
+## Rationale
+
+Broadening AWS platform support.
+
+Otherwise, users will need to run `telegraf` or another agent to ingest this
+data.
+
+## Prior Art
+
+* [telegraf](https://github.com/influxdata/telegraf/tree/master/plugins/inputs/cloudwatch)
+* [prometheus exporter](https://github.com/prometheus/cloudwatch_exporter)
+* [filebeat](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-cloudwatch.html)
+
+## Drawbacks
+
+The additional maintenance burden of a new source.
+
+## Alternatives
+
+There is alternate API,
+[`GetMetricStatistics'](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_GetMetricStatistics.html)
+that we could use, but I think it will result in less overall throughput due to
+API rate limit restrictions:
+
+`GetMetricStatistics`
+
+* Allows 400 transactions / sec
+* Each transaction can query the 5 standard metrics and up to 10 extended
+  metrics for one metric name / dimension combination
+* This should allow for: `400 * 5 = 2000` standard metrics / second
+
+`GetMetricData`
+
+* Allows 50 transactions / sec
+* Each transaction can query up to 500 metric name / metric stat (like
+  `average`) / dimension combination
+* This should allow for: `50 * 500 = 25000` standard metrics / second
+
+These are theoretical maxes. They could differ depending on how many metric
+statistics are queried for each metric. Querying less than the 5 standard ones
+would further advantage `GetMetricData`.
+
+## Outstanding Questions
+
+* Do we care about providing a different rate limit for refreshing available
+  metrics? It uses a different API (`ListMetrics`) with different limits than
+  `GetMetricData`. Answer: we'll try without for now.
+
+## Plan Of Attack
+
+* [ ] Submit PR with `aws_cloudwatch_metrics` source without any support for
+      globbing (which would require listing and caching metrics) and only one
+      region
+* [ ] Submit follow-up PR allowing for globbing of metric names
+* [ ] Submit follow-up PR allowing for globbing of dimension value
+* [ ] Submit follow-up PR allowing for globbing of namespace
+* [ ] Submit follow-up PR allowing for multiple regions
+
+## Future Work
+
+* Support for [metric
+  expressions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/using-metric-math.html#metric-math-syntax).
+* Filtering metrics by tags. It looks like the logstash input supports this by
+  querying for resources first (e.g. instances with a specific tag) and then
+  fetching metrics for those instances.
+* Backfilling metrics if `vector` is restarted or offline for a period.
+* Config examples covering common use-cases
```

Comment:
I assume we want at least _some_ use-cases covered as examples, before we stabilize the source?

Reply:
:+1: we should at least have a couple of examples in the documentation for the source. I'll update.